doc/api/stream.md

```diff
@@ -883,83 +1035,109 @@ implement Readable streams in your programs.
+
+When `readable._read()` is called, if data is available from the resource, the 
+implementation should begin pushing that data into the read queue using the
+[`this.push(dataChunk)`][stream-push] method. `_read()` should continue reading
+from the resource and pushing data until `readable.push()` returns `false`. Only
+when `_read()` is called again after it has stopped should it resume pushing
+additional data onto the queue.
+
+*Note*: Once the `readable._read()` method has been called, it will not be 
+called again until the [`readable.push()`][stream-push] method is called.
+
+The `size` argument is advisory. For implementations where a "read" is a
+single operation that returns data can use the `size` argument to determine how
+much data to fetch. Other implementations may ignore this argument and simply
+provide data whenever it becomes available. There is no need to "wait" until
 `size` bytes are available before calling [`stream.push(chunk)`][stream-push].
 
-#### readable.push(chunk[, encoding])
+The `readable.\_read()` method is prefixed with an underscore because it is
+internal to the class that defines it, and should never be called directly by
+user programs.
 
+#### readable.push(chunk[, encoding])
 
 * `chunk` {Buffer|Null|String} Chunk of data to push into the read queue
 * `encoding` {String} Encoding of String chunks.  Must be a valid
   Buffer encoding, such as `'utf8'` or `'ascii'`
 * return {Boolean} Whether or not more pushes should be performed
 
-Note: **This method should be called by Readable implementors, NOT
-by consumers of Readable streams.**
+*Note*: The `readable.push()` method is intended be called only by Readable
+implementors, and **not** by consumers of Readable streams.
 
-If a value other than null is passed, The `push()` method adds a chunk of data
-into the queue for subsequent stream processors to consume. If `null` is
-passed, it signals the end of the stream (EOF), after which no more data
-can be written.
+When `chunk` is a `Buffer` or `string`, the `chunk` of data will be added to the
+internal queue for users of the stream to consume. Passing `chunk` as `null` 
+signals the end of the stream (EOF), after which no more data can be written.
 
-The data added with `push()` can be pulled out by calling the
-[`stream.read()`][stream-read] method when the [`'readable'`][] event fires.
+When the Readable is operating in paused mode, the data added with
+`readable.push()` can be read out by calling the [`stream.read()`][stream-read]
+method when the [`'readable'`][] event fires.
 
-This API is designed to be as flexible as possible. For example,
-you may be wrapping a lower-level source which has some sort of
-pause/resume mechanism, and a data callback. In those cases, you
-could wrap the low-level source object by doing something like this:
+When the Readable is operating in flowing mode, the data added with
+`readable.push()` will be delivered by emitting a `'data'` event.
```

Comment:
There's a disconnect between `.push()` and transferring the data out of the stream (via `"data"` or `.read()`). In particular, `Readable` streams have a cushion similar to `Writable` streams — again, intended to downsample backpressure so that we don't flip our underlying resource between "reading" and "not reading" quite as often. In particular, even after you've `push`'d data, if it wasn't _enough_ data to fill the hwm, `_read` will be called again, and again, and again, until the internal buffer has filled to hwm. The net effect is to delay backpressure from reaching the underlying resource, in the hopes that the backpressure will drain before we have to stop the underlying resource.